File: python/239.py
# Sliding Window Maximum
# 滑动窗口最大值


class Solution:
    def maxSlidingWindow(self, nums: list[int], k: int) -> list[int]:
        answer = []
        heap = [(-num, i) for i, num in enumerate(nums[:k])]
        import heapq
        heapq.heapify(heap)
        answer.append(-heap[0][0])

        for i in range(k, len(nums)):
            heapq.heappush(heap, (-nums[i], i))
            while heap[0][1] <= i - k:
                heapq.heappop(heap)
            answer.append(-heap[0][0])
        return answer

if __name__ == '__main__':
    test_cases = [
        ([1,3,-1,-3,5,3,6,7], 3),
        ([1], 1),
        ([1,-1], 1),
        ([9,11], 2),
        ([4,-2], 2),
    ]

    sol = Solution()
    for case in test_cases:
        res = sol.maxSlidingWindow(*case)
        print(case, '\n', res)

File: cpp/239.cpp
// 滑动窗口最大值
// Sliding Window Maximum

#include "common.h"

class Solution
{
public:
    std::vector<int> maxSlidingWindow(std::vector<int> &nums, int k)
    {
        std::priority_queue<std::pair<int, int>> max_heap;
        for (int i = 0; i < k; ++i)
        {
            max_heap.emplace(nums[i], i);
        }
        std::vector<int> result = {max_heap.top().first};

        for (int i = k; i < nums.size(); ++i)
        {
            max_heap.emplace(nums[i], i);
            while (max_heap.top().second <= i - k)
            {
                max_heap.pop();
            }
            result.push_back(max_heap.top().first);
        }
        return result;
    }
};

int main()
{
    std::vector<std::tuple<std::vector<int>, int>> test_cases = {
        {{1, 3, -1, -3, 5, 3, 6, 7}, 3},
        {{1}, 1},
    };

    for (auto &[nums, k] : test_cases)
    {
        print_vector(Solution().maxSlidingWindow(nums, k));
    }
    return 0;
}